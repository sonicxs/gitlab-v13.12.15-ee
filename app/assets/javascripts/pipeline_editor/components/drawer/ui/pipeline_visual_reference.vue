<script>
import { s__ } from '~/locale';
import DemoJobPill from './demo_job_pill.vue';

export default {
  i18n: {
    stageNames: {
      build: s__('StageName|Build'),
      test: s__('StageName|Test'),
      deploy: s__('StageName|Deploy'),
    },
    jobNames: {
      build: s__('JobName|build-job'),
      test_1: s__('JobName|unit-test'),
      test_2: s__('JobName|lint-test'),
      deploy: s__('JobName|deploy-app'),
    },
  },
  stageClasses:
    'gl-bg-blue-50 gl-display-flex gl-flex-direction-column gl-align-items-center gl-p-4 gl-rounded-base',
  titleClasses: 'gl-text-blue-600 gl-mb-4',
  components: {
    DemoJobPill,
  },
};
</script>
<template>
  <div class="gl-display-flex gl-justify-content-center">
    <div :class="$options.stageClasses" class="gl-mr-5">
      <div :class="$options.titleClasses">{{ $options.i18n.stageNames.build }}</div>
      <demo-job-pill :job-name="$options.i18n.jobNames.build" />
    </div>
    <div :class="$options.stageClasses" class="gl-mr-5">
      <div :class="$options.titleClasses">{{ $options.i18n.stageNames.test }}</div>
      <demo-job-pill class="gl-mb-3" :job-name="$options.i18n.jobNames.test_1" />
      <demo-job-pill :job-name="$options.i18n.jobNames.test_2" />
    </div>
    <div :class="$options.stageClasses">
      <div :class="$options.titleClasses">{{ $options.i18n.stageNames.deploy }}</div>
      <demo-job-pill :job-name="$options.i18n.jobNames.deploy" />
    </div>
  </div>
</template>
