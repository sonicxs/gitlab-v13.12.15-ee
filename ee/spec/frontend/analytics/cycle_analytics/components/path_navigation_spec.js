import { GlPath, GlDeprecatedSkeletonLoading as GlSkeletonLoading } from '@gitlab/ui';
import { mount } from '@vue/test-utils';
import Component from 'ee/analytics/cycle_analytics/components/path_navigation.vue';
import { transformedStagePathData, issueStage } from '../mock_data';

describe('PathNavigation', () => {
  let wrapper = null;

  const createComponent = (props) => {
    return mount(Component, {
      propsData: {
        stages: transformedStagePathData,
        selectedStage: issueStage,
        loading: false,
        ...props,
      },
    });
  };

  const pathNavigationTitles = () => {
    return wrapper.findAll('.gl-path-button');
  };

  const pathNavigationItems = () => {
    return wrapper.findAll('.gl-path-nav-list-item');
  };

  const clickItemAt = (index) => {
    pathNavigationTitles().at(index).trigger('click');
  };

  beforeEach(() => {
    wrapper = createComponent();
  });

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  describe('displays correctly', () => {
    it('has the correct props', () => {
      expect(wrapper.find(GlPath).props('items')).toMatchObject(transformedStagePathData);
    });

    it('contains all the expected stages', () => {
      const html = wrapper.find(GlPath).html();

      transformedStagePathData.forEach((stage) => {
        expect(html).toContain(stage.title);
      });
    });

    describe('loading', () => {
      describe('is false', () => {
        it('displays the gl-path component', () => {
          expect(wrapper.find(GlPath).exists()).toBe(true);
        });

        it('hides the gl-skeleton-loading component', () => {
          expect(wrapper.find(GlSkeletonLoading).exists()).toBe(false);
        });

        it('matches the snapshot', () => {
          expect(wrapper.element).toMatchSnapshot();
        });

        describe('popovers', () => {
          beforeEach(() => {
            wrapper = createComponent({ stages: transformedStagePathData });
          });

          it('renders popovers for all stages except for the overview stage', () => {
            const pathItemContent = pathNavigationItems().wrappers;
            const [overviewStage, ...popoverStages] = pathItemContent;

            expect(overviewStage.text()).toContain('Overview');
            expect(overviewStage.find('[data-testid="stage-item-popover"]').exists()).toBe(false);

            popoverStages.forEach((stage) => {
              expect(stage.find('[data-testid="stage-item-popover"]').exists()).toBe(true);
            });
          });

          it('shows the sanitized start event description for the first stage item', () => {
            const firstPopover = wrapper.findAll('[data-testid="stage-item-popover"]').at(0);
            const expectedStartEventDescription = 'Issue created';
            expect(firstPopover.text()).toContain(expectedStartEventDescription);
          });

          it('shows the sanitized end event description for the first stage item', () => {
            const firstPopover = wrapper.findAll('[data-testid="stage-item-popover"]').at(0);
            const expectedStartEventDescription =
              'Issue first associated with a milestone or issue first added to a board';
            expect(firstPopover.text()).toContain(expectedStartEventDescription);
          });

          it('shows the median stage time for the first stage item', () => {
            const firstPopover = wrapper.findAll('[data-testid="stage-item-popover"]').at(0);
            expect(firstPopover.text()).toContain('Stage time (median)');
          });
        });
      });

      describe('is true', () => {
        beforeEach(() => {
          wrapper = createComponent({ loading: true });
        });

        it('hides the gl-path component', () => {
          expect(wrapper.find(GlPath).exists()).toBe(false);
        });

        it('displays the gl-skeleton-loading component', () => {
          expect(wrapper.find(GlSkeletonLoading).exists()).toBe(true);
        });
      });
    });
  });

  describe('event handling', () => {
    it('emits the selected event', () => {
      expect(wrapper.emitted('selected')).toBeUndefined();

      clickItemAt(0);
      clickItemAt(1);
      clickItemAt(2);

      expect(wrapper.emitted().selected).toEqual([
        [transformedStagePathData[0]],
        [transformedStagePathData[1]],
        [transformedStagePathData[2]],
      ]);
    });
  });
});
