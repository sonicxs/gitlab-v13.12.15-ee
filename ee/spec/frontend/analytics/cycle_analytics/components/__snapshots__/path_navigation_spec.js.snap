// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathNavigation displays correctly loading is false matches the snapshot 1`] = `
<div
  class="gl-path-nav"
  data-testid="gl-path-nav"
  style="--path-bg-color: white;"
>
  <span
    class="gl-path-fade gl-path-fade-left"
    style="display: none;"
  >
    <button
      aria-label="Scroll left"
      class="gl-clear-icon-button"
    >
      <svg
        aria-hidden="true"
        class="gl-icon s32"
        data-testid="chevron-left-icon"
      >
        <use
          href="#chevron-left"
        />
      </svg>
    </button>
  </span>
   
  <ul
    class="gl-path-nav-list"
  >
    <li
      class="gl-path-nav-list-item"
      id="path-6-item-0"
    >
      <button
        class="gl-path-button"
      >
        <svg
          aria-hidden="true"
          class="gl-mr-2 gl-icon s16"
          data-testid="gl-path-item-icon"
        >
          <use
            href="#home"
          />
        </svg>
        
        Overview
        <!---->
      </button>
       
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-6-item-1"
    >
      <button
        class="gl-path-button gl-path-active-item-indigo"
      >
        <!---->
        
        Issue
        <!---->
      </button>
       
      <div
        class="gl-popover"
        data-testid="stage-item-popover"
      >
         
        <div
          class="gl-px-4"
        >
          <div
            class="gl-display-flex gl-justify-content-space-between"
          >
            <div
              class="gl-pr-4 gl-pb-4"
            >
              
            Stage time (median)
          
            </div>
             
            <div
              class="gl-pb-4 gl-font-weight-bold"
            >
              
            </div>
          </div>
        </div>
         
        <div
          class="gl-px-4"
        >
          <div
            class="gl-display-flex gl-justify-content-space-between"
          >
            <div
              class="gl-pr-4 gl-pb-4"
            >
              
            Items in stage
          
            </div>
             
            <div
              class="gl-pb-4 gl-font-weight-bold"
            >
              172800 items
            </div>
          </div>
        </div>
         
        <div
          class="gl-px-4 gl-pt-4 gl-border-t-1 gl-border-t-solid gl-border-gray-50"
        >
          <div
            class="gl-display-flex gl-flex-direction-row"
          >
            <div
              class="gl-display-flex gl-flex-direction-column gl-pr-4 gl-pb-4 metric-label"
            >
              
            Start
          
            </div>
             
            <div
              class="gl-display-flex gl-flex-direction-column gl-pb-4 stage-event-description"
            >
              <p
                data-sourcepos="1:1-1:13"
                dir="auto"
              >
                Issue created
              </p>
            </div>
          </div>
           
          <div
            class="gl-display-flex gl-flex-direction-row"
          >
            <div
              class="gl-display-flex gl-flex-direction-column gl-pr-4 metric-label"
            >
              
            Stop
          
            </div>
             
            <div
              class="gl-display-flex gl-flex-direction-column stage-event-description"
            >
              <p
                data-sourcepos="1:1-1:71"
                dir="auto"
              >
                Issue first associated with a milestone or issue first added to a board
              </p>
            </div>
          </div>
        </div>
        Issue
      </div>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-6-item-2"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Plan
        <!---->
      </button>
       
      <div
        class="gl-popover"
        data-testid="stage-item-popover"
      >
         
        <div
          class="gl-px-4"
        >
          <div
            class="gl-display-flex gl-justify-content-space-between"
          >
            <div
              class="gl-pr-4 gl-pb-4"
            >
              
            Stage time (median)
          
            </div>
             
            <div
              class="gl-pb-4 gl-font-weight-bold"
            >
              
            </div>
          </div>
        </div>
         
        <div
          class="gl-px-4"
        >
          <div
            class="gl-display-flex gl-justify-content-space-between"
          >
            <div
              class="gl-pr-4 gl-pb-4"
            >
              
            Items in stage
          
            </div>
             
            <div
              class="gl-pb-4 gl-font-weight-bold"
            >
              86400 items
            </div>
          </div>
        </div>
         
        <div
          class="gl-px-4 gl-pt-4 gl-border-t-1 gl-border-t-solid gl-border-gray-50"
        >
          <div
            class="gl-display-flex gl-flex-direction-row"
          >
            <div
              class="gl-display-flex gl-flex-direction-column gl-pr-4 gl-pb-4 metric-label"
            >
              
            Start
          
            </div>
             
            <div
              class="gl-display-flex gl-flex-direction-column gl-pb-4 stage-event-description"
            >
              <p
                data-sourcepos="1:1-1:71"
                dir="auto"
              >
                Issue first associated with a milestone or issue first added to a board
              </p>
            </div>
          </div>
           
          <div
            class="gl-display-flex gl-flex-direction-row"
          >
            <div
              class="gl-display-flex gl-flex-direction-column gl-pr-4 metric-label"
            >
              
            Stop
          
            </div>
             
            <div
              class="gl-display-flex gl-flex-direction-column stage-event-description"
            >
              <p
                data-sourcepos="1:1-1:33"
                dir="auto"
              >
                Issue first mentioned in a commit
              </p>
            </div>
          </div>
        </div>
        Plan
      </div>
    </li>
    <li
      class="gl-path-nav-list-item"
      id="path-6-item-3"
    >
      <button
        class="gl-path-button"
      >
        <!---->
        
        Code
        <!---->
      </button>
       
      <div
        class="gl-popover"
        data-testid="stage-item-popover"
      >
         
        <div
          class="gl-px-4"
        >
          <div
            class="gl-display-flex gl-justify-content-space-between"
          >
            <div
              class="gl-pr-4 gl-pb-4"
            >
              
            Stage time (median)
          
            </div>
             
            <div
              class="gl-pb-4 gl-font-weight-bold"
            >
              
            </div>
          </div>
        </div>
         
        <div
          class="gl-px-4"
        >
          <div
            class="gl-display-flex gl-justify-content-space-between"
          >
            <div
              class="gl-pr-4 gl-pb-4"
            >
              
            Items in stage
          
            </div>
             
            <div
              class="gl-pb-4 gl-font-weight-bold"
            >
              129600 items
            </div>
          </div>
        </div>
         
        <div
          class="gl-px-4 gl-pt-4 gl-border-t-1 gl-border-t-solid gl-border-gray-50"
        >
          <div
            class="gl-display-flex gl-flex-direction-row"
          >
            <div
              class="gl-display-flex gl-flex-direction-column gl-pr-4 gl-pb-4 metric-label"
            >
              
            Start
          
            </div>
             
            <div
              class="gl-display-flex gl-flex-direction-column gl-pb-4 stage-event-description"
            >
              <p
                data-sourcepos="1:1-1:33"
                dir="auto"
              >
                Issue first mentioned in a commit
              </p>
            </div>
          </div>
           
          <div
            class="gl-display-flex gl-flex-direction-row"
          >
            <div
              class="gl-display-flex gl-flex-direction-column gl-pr-4 metric-label"
            >
              
            Stop
          
            </div>
             
            <div
              class="gl-display-flex gl-flex-direction-column stage-event-description"
            >
              <p
                data-sourcepos="1:1-1:21"
                dir="auto"
              >
                Merge request created
              </p>
            </div>
          </div>
        </div>
        Code
      </div>
    </li>
  </ul>
   
  <span
    class="gl-path-fade gl-path-fade-right"
    style="display: none;"
  >
    <button
      aria-label="Scroll right"
      class="gl-clear-icon-button"
    >
      <svg
        aria-hidden="true"
        class="gl-icon s32"
        data-testid="chevron-right-icon"
      >
        <use
          href="#chevron-right"
        />
      </svg>
    </button>
  </span>
</div>
`;
